### PR TITLE
Add test_mpp_transmit

### PR DESCRIPTION
**Pull request by laurenchilutti**
**Description**
Adding test_mpp_transmit which supports r4, r8, i4, and i8
Also updating test_mpp_sum to use platform_mod.

Fixes # (issue)

**How Has This Been Tested?**
Tested on Gaea using compilers loaded with fre/bronx-18

**Checklist:**
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] Any dependent changes have been merged and published in downstream modules
- [x] New check tests, if applicable, are included
- [x] `make distcheck` passes

### CODE REVIEW

File: test_fms/mpp/test_mpp_transmit.F90 [hunk 1]
```diff
@@ -0,0 +1,532 @@
...
+    real(kind=i8_kind), dimension(1,1) :: a8, b8, c8
+
+    n=1
+    a8 = real(pe, kind=i8_kind)
+    call mpp_sync()
+
+    call mpp_transmit( put_data=a8(1,1), plen=n, to_pe=modulo(pe+1, npes), &
+                       get_data=b8(1,1), glen=n, from_pe=modulo(npes+pe-1, npes) )
+    call mpp_sync_self()
+    c8 = real(modulo(int(a8(1,1))+npes-1,npes), kind=i8_kind)
```
Comment (wrongkindofdoctor) [inline test_fms/mpp/test_mpp_transmit.F90:266]: `modulo(a8(1,1)+npes-1,npes)`?

File: test_fms/mpp/test_mpp_transmit.F90 [hunk 2]
```diff
@@ -0,0 +1,532 @@
...
+  end subroutine test_mpp_transmit_i4_2D
+
+  !> Test the functionality of mpp_transmit for an i8_2D.
+  subroutine test_mpp_transmit_i8_2D(npes,pe,root)
+    integer, intent(in) :: npes, pe, root
+    integer :: n
+    real(kind=i8_kind), dimension(1,1) :: a8, b8, c8
+
+    n=1
+    a8 = real(pe, kind=i8_kind)
```
Comment (wrongkindofdoctor) [inline test_fms/mpp/test_mpp_transmit.F90:260]: `int(pe, kind=i8_kind)`

File: test_fms/mpp/test_mpp_transmit.F90 [hunk 3]
```diff
@@ -0,0 +1,532 @@
...
+    ! b4(1,1) should now equal the value of a4(1,1)from the "from_pe"
+    if (b4(1,1) .NE. c4(1,1) ) call mpp_error(FATAL, "Test_mpp_transmit_i4_2D: transmit didn't go as expected")
+
+  end subroutine test_mpp_transmit_i4_2D
+
+  !> Test the functionality of mpp_transmit for an i8_2D.
+  subroutine test_mpp_transmit_i8_2D(npes,pe,root)
+    integer, intent(in) :: npes, pe, root
+    integer :: n
+    real(kind=i8_kind), dimension(1,1) :: a8, b8, c8
```
Comment (wrongkindofdoctor) [inline test_fms/mpp/test_mpp_transmit.F90:257]: `int(kind=i8_kind)`

File: test_fms/mpp/test_mpp_transmit.F90 [hunk 4]
```diff
@@ -0,0 +1,532 @@
...
+    real(kind=i4_kind), dimension(1,1) :: a4, b4, c4
+
+    n=1
+    a4 = real(pe, kind=i4_kind)
+    call mpp_sync()
+
+    call mpp_transmit( put_data=a4(1,1), plen=n, to_pe=modulo(pe+1, npes), &
+                       get_data=b4(1,1), glen=n, from_pe=modulo(npes+pe-1, npes) )
+    call mpp_sync_self()
+    c4 = real(modulo(int(a4(1,1))+npes-1,npes), kind=i4_kind)
```
Comment (wrongkindofdoctor) [inline test_fms/mpp/test_mpp_transmit.F90:247]: `modulo(a4(1,1)+npes-1,npes)`

File: test_fms/mpp/test_mpp_transmit.F90 [hunk 5]
```diff
@@ -0,0 +1,532 @@
...
+  end subroutine test_mpp_transmit_r8_2D
+
+  !> Test the functionality of mpp_transmit for an i4_2D.
+  subroutine test_mpp_transmit_i4_2D(npes,pe,root)
+    integer, intent(in) :: npes, pe, root
+    integer :: n
+    real(kind=i4_kind), dimension(1,1) :: a4, b4, c4
+
+    n=1
+    a4 = real(pe, kind=i4_kind)
```
Comment (wrongkindofdoctor) [inline test_fms/mpp/test_mpp_transmit.F90:241]: `int(pe,kind=i4_kind)`

File: test_fms/mpp/test_mpp_transmit.F90 [hunk 6]
```diff
@@ -0,0 +1,532 @@
...
+    ! b8(1,1) should now equal the value of a8(1,1)from the "from_pe"
+    if (b8(1,1) .NE. c8(1,1) ) call mpp_error(FATAL, "Test_mpp_transmit_r8_2D: transmit didn't go as expected")
+
+  end subroutine test_mpp_transmit_r8_2D
+
+  !> Test the functionality of mpp_transmit for an i4_2D.
+  subroutine test_mpp_transmit_i4_2D(npes,pe,root)
+    integer, intent(in) :: npes, pe, root
+    integer :: n
+    real(kind=i4_kind), dimension(1,1) :: a4, b4, c4
```
Comment (wrongkindofdoctor) [inline test_fms/mpp/test_mpp_transmit.F90:238]: `int(kind=i4_kind)`

File: test_fms/mpp/test_mpp_transmit.F90 [hunk 7]
```diff
@@ -0,0 +1,532 @@
...
+    real(kind=i8_kind), dimension(1) :: a8, b8, c8
+
+    n=1
+    a8 = real(pe, kind=i8_kind)
+    call mpp_sync()
+
+    call mpp_transmit( put_data=a8(1), plen=n, to_pe=modulo(pe+1, npes), &
+                       get_data=b8(1), glen=n, from_pe=modulo(npes+pe-1, npes) )
+    call mpp_sync_self()
+    c8 = real(modulo(int(a8(1))+npes-1,npes), kind=i8_kind)
```
Comment (wrongkindofdoctor) [inline test_fms/mpp/test_mpp_transmit.F90:180]: `modulo(a8(1)+npes-1,npes)`

File: test_fms/mpp/test_mpp_transmit.F90 [hunk 8]
```diff
@@ -0,0 +1,532 @@
...
+    ! b4(1) should now equal the value of a4(1)from the "from_pe"
+    if (b4(1) .NE. c4(1) ) call mpp_error(FATAL, "Test_mpp_transmit_i4_scalar: transmit didn't go as expected")
+
+  end subroutine test_mpp_transmit_i4_scalar
+
+  !> Test the functionality of mpp_transmit for an i8_scalar.
+  subroutine test_mpp_transmit_i8_scalar(npes,pe,root)
+    integer, intent(in) :: npes, pe, root
+    integer :: i, n
+    real(kind=i8_kind), dimension(1) :: a8, b8, c8
```
Comment (wrongkindofdoctor) [inline test_fms/mpp/test_mpp_transmit.F90:171]: `int(kind=i8_kind)`

File: test_fms/mpp/test_mpp_transmit.F90 [hunk 9]
```diff
@@ -0,0 +1,532 @@
...
+
+    n=1
+    a8 = real(pe, kind=i8_kind)
```
Comment (wrongkindofdoctor) [inline test_fms/mpp/test_mpp_transmit.F90:174]: `int(pe, kind=i8_kind)`

File: test_fms/mpp/test_mpp_transmit.F90 [hunk 10]
```diff
@@ -0,0 +1,532 @@
...
+    real(kind=i4_kind), dimension(1) :: a4, b4, c4
+
+    n=1
+    a4 = real(pe, kind=i4_kind)
+    call mpp_sync()
+
+    call mpp_transmit( put_data=a4(1), plen=n, to_pe=modulo(pe+1, npes), &
+                       get_data=b4(1), glen=n, from_pe=modulo(npes+pe-1, npes) )
+    call mpp_sync_self()
+    c4 = real(modulo(int(a4(1))+npes-1,npes), kind=i4_kind)
```
Comment (wrongkindofdoctor) [inline test_fms/mpp/test_mpp_transmit.F90:161]: `modulo(a4(1)+npes-1,npes)`

File: test_fms/mpp/test_mpp_transmit.F90 [hunk 11]
```diff
@@ -0,0 +1,532 @@
...
+    ! b8(1) should now equal the value of a8(1)from the "from_pe"
+    if (b8(1) .NE. c8(1) ) call mpp_error(FATAL, "Test_mpp_transmit_r8_scalar: transmit didn't go as expected")
+
+  end subroutine test_mpp_transmit_r8_scalar
+
+  !> Test the functionality of mpp_transmit for an i4_scalar.
+  subroutine test_mpp_transmit_i4_scalar(npes,pe,root)
+    integer, intent(in) :: npes, pe, root
+    integer :: n
+    real(kind=i4_kind), dimension(1) :: a4, b4, c4
```
Comment (wrongkindofdoctor) [inline test_fms/mpp/test_mpp_transmit.F90:152]: `int(kind=i4_kind)`

File: test_fms/mpp/test_mpp_transmit.F90 [hunk 12]
```diff
@@ -0,0 +1,532 @@
...
+
+    n=1
+    a4 = real(pe, kind=i4_kind)
```
Comment (wrongkindofdoctor) [inline test_fms/mpp/test_mpp_transmit.F90:155]: `int(pe, kind=i4_kind)`

**Review by wrongkindofdoctor — CHANGES_REQUESTED** [review]

**Review by ngs333 — APPROVED** [review]

**Review by wrongkindofdoctor — APPROVED** [review]

**Review by mlee03 — COMMENTED** [review]
mpp_init accepts several optional arguments, should use mpp_init(test_level=mpp_init_test_requests_allocated) ?

**Comment by laurenchilutti** [comment]
@mlee03 I updated the mpp_init call to match what you've said.

File: test_fms/mpp/test_mpp_transmit.F90 [hunk 13]
```diff
@@ -0,0 +1,532 @@
...
+
+  !> Test the functionality of mpp_transmit for an r4_scalar.
+  subroutine test_mpp_transmit_r4_scalar(npes,pe,root)
+    integer, intent(in) :: npes, pe, root
+    integer :: n
+    real(kind=r4_kind), dimension(1) :: a4, b4, c4
+
+    n=1
+    a4 = real(pe, kind=r4_kind)
+    call mpp_sync()
```
Comment (mlee03) [inline test_fms/mpp/test_mpp_transmit.F90:117]: not sure if it's necessary or not, but b4 is not initialized

File: test_fms/mpp/test_mpp_transmit.F90 [hunk 14]
```diff
@@ -0,0 +1,532 @@
...
+  subroutine test_mpp_transmit_r4_2D(npes,pe,root)
+    integer, intent(in) :: npes, pe, root
+    integer :: i, n
+    real(kind=r4_kind), dimension(1,1) :: a4, b4, c4
+
+    n=1
+    a4 = real(pe, kind=r4_kind)
+    call mpp_sync()
+
+    call mpp_transmit( put_data=a4(1,1), plen=n, to_pe=modulo(pe+1, npes), &
```
Comment (mlee03) [inline test_fms/mpp/test_mpp_transmit.F90:206]: is it supposed to be put_data=a4?

File: test_fms/mpp/test_mpp_transmit.F90 [hunk 15]
```diff
@@ -0,0 +1,532 @@
...
+    ! b4(1,1) should now equal the value of a4(1,1)from the "from_pe"
+    if (b4(1,1) .NE. c4(1,1) ) call mpp_error(FATAL, "Test_mpp_transmit_r4_2D: transmit didn't go as expected")
+
+  end subroutine test_mpp_transmit_r4_2D
+
+  !> Test the functionality of mpp_transmit for an r8_2D.
+  subroutine test_mpp_transmit_r8_2D(npes,pe,root)
+    integer, intent(in) :: npes, pe, root
+    integer :: n
+    real(kind=r8_kind), dimension(1,1) :: a8, b8, c8
```
Comment (mlee03) [inline test_fms/mpp/test_mpp_transmit.F90:219]: perhaps something bigger than (1,1)?  I think for (1,1), the size of the array is just 1

File: test_fms/mpp/test_mpp_transmit.F90 (hunk 13 above)
Comment (wrongkindofdoctor) [reply test_fms/mpp/test_mpp_transmit.F90:117]: Yes, it would be good to initialize the arrays.

Comment (laurenchilutti) [reply test_fms/mpp/test_mpp_transmit.F90:119]: added this in the latest commit

File: test_fms/mpp/test_mpp_transmit.F90 (hunk 14 above)
Comment (laurenchilutti) [reply test_fms/mpp/test_mpp_transmit.F90]: From testing, this does not work when put_data=a4.  In [mpp_transmit](https://github.com/NOAA-GFDL/FMS/blob/master/mpp/include/mpp_transmit_mpi.h#L81), put_data  is being used as the "buff" input parameter to [MPI_ISEND](https://www.mpich.org/static/docs/latest/www3/MPI_Isend.html).  The "buff" parameter is defined as "initial address of send buffer".  I guess a4(1,1) is the initial address of the send buffer.

File: test_fms/mpp/test_mpp_transmit.F90 (hunk 15 above)
Comment (laurenchilutti) [reply test_fms/mpp/test_mpp_transmit.F90]: Good point, I have updated to have dimensions in increments of 2 so we can truly test the capability.

**Pull request merged**